### ISSUE TITLE
net: `ParseIP` returns nil instead of IP to a seemingly valid IPv6

### ISSUE THREAD

**Issue by ucirello** [golang/go#69752]
### Go version

go1.23.2

### Output of `go env` in your module/workspace:

```shell
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/USER/Library/Caches/go-build'
GOENV='/Users/USER/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/USER/go/pkg/mod/'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/USER/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.23.2'
GODEBUG=''
GOTELEMETRY='on'
GOTELEMETRYDIR='/Users/USER/Library/Application Support/go/telemetry'
GCCGO='gccgo'
GOARM64='v8.0'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOMOD='/Users/USER/go/src/github.com/axiomhq/axiom/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/rb/xq45klk13kg60s3mqqr0ff7c0000gn/T/go-build1766105564=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

### What did you do?

I parsed an IP using this code:

```
package main

import (
	"fmt"
	"net"
)

func main() {
	fmt.Println(net.ParseIP("0:0:0:0:000000:ffff:127.5.8.7").To4())
	fmt.Printf("%#v", net.ParseIP("0:0:0:0:000000:ffff:127.5.8.7"))
}
```

### What did you see happen?

https://go.dev/play/p/bJUjbkLm_bg (Go 1.23)

```
<nil>
net.IP(nil)
```

### What did you expect to see?

https://go.dev/play/p/bJUjbkLm_bg?v=goprev  (Go 1.22)

```
127.5.8.7
net.IP{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 0x7f, 0x5, 0x8, 0x7}
```

**Comment by seankhliao** [comment]
That's not valid, there's too many digits in `000000`

**Comment by ucirello** [comment]
According to https://datatracker.ietf.org/doc/html/rfc4291#section-2.2

>  1. The preferred form is x:x:x:x:x:x:x:x, where the 'x's are one to
      four hexadecimal digits of the eight 16-bit pieces of the address.

Meaning that the example at hand is invalid. 

I don't thinkg this behavior change is covered by Go 1 Compatibility Promise, correct?

**Comment by mknyszek** [comment]
https://go.dev/cl/461605 is the change that added this behavior, which references #57760.

So, this change does seem deliberate. I'm not sure what the right path forward is here.

CC @neild @ianlancetaylor who reviewed the CL.

**Comment by mknyszek** [comment]
Reopening for now to await input from Damien and/or Ian.

**Comment by ianlancetaylor** [comment]
CC @mpetrunic

**Comment by neild** [comment]
Unless I'm missing something, `0:0:0:0:000000:ffff:127.5.8.7` is not a valid IPv6 address. As @seankhliao says, there are too many zeros.

Go 1.22 and prior had a bug which caused us to parse this address anyway. (#57760)

The bug was fixed in Go 1.23.

This isn't a violation of the compatibility promise, since we reserve the right to fix bugs. `net.ParseIP`'s documentation says that it returns nil if given an invalid input, so this is definitely a bug. (Unless we're all very confused and this address is actually valid.)

I don't think there's anything for us to do here.
